### PR TITLE
Starting HubOptions

### PR DESCRIPTION
**Pull request by BrennanConroy**
https://github.com/aspnet/SignalR/issues/254 and https://github.com/aspnet/SignalR/issues/261

Wanted to get feedback to see if this is the design we want. Then I'll add a test.
Do we want it to be `HubOptions<THub>`? 

Current usage:
```c#
services.Configure<HubOptions>(options =>
{
    options.JsonSerializerSettings = new JsonSerializerSettings()
    {
        ContractResolver = new CamelCasePropertyNamesContractResolver()
    };
});
```

### CODE REVIEW

**Comment by davidfowl** [comment]
This should also work when I call AddSignalR

File: src/Microsoft.AspNetCore.SignalR/Internal/DefaultHubProtocolResolver.cs [hunk 1]
```diff
@@ -3,18 +3,26 @@
...
+            _options = options;
+        }
+
         public IHubProtocol GetProtocol(string protocolName, HubConnectionContext connection)
         {
             switch (protocolName?.ToLowerInvariant())
             {
                 case "json":
-                    return new JsonHubProtocol(new JsonSerializer());
+                    return new JsonHubProtocol(JsonSerializer.Create(_options.Value.JsonSerializerSettings));
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR/Internal/DefaultHubProtocolResolver.cs]: what settings are used if JsonSerializerSettings is null?

Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR/Internal/DefaultHubProtocolResolver.cs]: The default ones. Exact same as what we were doing before.

**Comment by moozzyk** [comment]
I don't think we can JSON settings defined by the user to serialize/read our payload

**Comment by BrennanConroy** [comment]
It doesn't change any of our items. Example:
```
await Clients.All.InvokeAsync("Send", $"{Context.ConnectionId} joined", new { Message = "m", TestName = "t" });
```
Sends:
```
data received: {"invocationId":"1","type":1,"target":"Send","nonBlocking":true,"arguments":["735fe372-7d10-4a6a-9e69-35d3ece80e85 joined",{"message":"m","testName":"t"}]}
```

We use `JsonTextWriter` just talked with @moozzyk in person

**Comment by davidfowl** [comment]
Where's the test?

File: src/Microsoft.AspNetCore.SignalR.Http/SignalRDependencyInjectionExtensions.cs [hunk 2]
```diff
@@ -9,6 +10,12 @@ public static class SignalRDependencyInjectionExtensions
     {
         public static ISignalRBuilder AddSignalR(this IServiceCollection services)
         {
+            return AddSignalR(services, o => { });
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Http/SignalRDependencyInjectionExtensions.cs]: _ => { }

**Comment by BrennanConroy** [comment]
🆙 📅

**Comment by davidfowl** [comment]
@moozzyk I think so, but we can revisit that after alpha.

File: src/Microsoft.AspNetCore.SignalR/HubEndPoint.cs [hunk 3]
```diff
@@ -36,16 +37,19 @@ public class HubEndPoint<THub> : IInvocationBinder where THub : Hub
         private readonly ILogger<HubEndPoint<THub>> _logger;
         private readonly IServiceScopeFactory _serviceScopeFactory;
         private readonly IHubProtocolResolver _protocolResolver;
+        private readonly IOptions<HubOptions> _hubOptions;
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR/HubEndPoint.cs]: Do we have any concrete things to add here for alpha?

**Comment by BrennanConroy** [comment]
Original issue that has a couple items, none look too important https://github.com/aspnet/SignalR/issues/254

**Pull request merged**